### PR TITLE
[Sampling] Support sampling masks with overlap scheduling

### PR DESCRIPTION
**Pull request by ByronHsu**
## Motivation

Sampling masks currently do not support overlap scheduling because materializing the masks inside the sampler synchronizes the GPU with the CPU.

## Modifications

1. Remove CPU synchronization from sampling-mask capture. Keep sampling-mask outputs on the GPU inside the sampler, then copy them asynchronously to the CPU and materialize the response data during result processing.
2. Introduce `--sampling-mask-max-tokens` (default: `4096`) to set the maximum number of tokens per sampling mask.
3. Reject sampling-mask requests upfront if `top_k` exceeds the configured mask capacity.
4. Abort the request with an overflow error if the realized mask exceeds the capacity, including when cutoff ties produce more tokens than `top_k` even though `top_k` is within the limit.
5. Mark sampling results as invalid if validation fails, such as when the selected-token logprob is non-finite (`NaN` or infinity). Abort the request before committing the failed step's sampled token.
6. Replace `SGLANG_DISAGGREGATION_SAMPLING_MASK_MAX_TOKENS` with the server argument `--sampling-mask-max-tokens` for both ordinary and disaggregated serving. PD metadata remains off by default: set `SGLANG_ENABLE_DISAGG_SAMPLING_MASK=1` on both prefill and decode servers. Without it, requests asking for masks receive HTTP 400 with the required setting; unused metadata buffers are neither allocated nor transferred. Startup rejects the retired `SGLANG_DISAGGREGATION_SAMPLING_MASK_MAX_TOKENS` variable with explicit migration instructions.

## Accuracy Tests

Current head `d73bc1bd7f`; review fixes verified at `1d542447e5` on BJ job `3439034` (NVIDIA H200, `lmsysorg/sglang:dev-cu13`):

- 240 affected unit tests and 35 subtests passed. Before-fix tests reproduced premature KV slot reuse during offload, lost error types/cleanup interruption on prefill abort, and all eight legacy-env configurations silently starting.
- Live TP2/PP1 and TP1/PP2 sampling-mask tests passed with `Qwen/Qwen2.5-0.5B-Instruct`: masks with and without ordinary logprobs, followed by ordinary generation; four tokens per request. Checks cover selected-token membership, support IDs, aligned lengths, capacity, and finite sampling logprobs.
- TP2 uses overlap and the real status collective. PP2 uses the existing pipeline scheduler (which automatically disables ordinary overlap) and real cross-stage send/receive.
- After consolidating the distributed cases into the existing sampling-mask file, all 29 tests passed in 249.286 seconds on BJ with a runtime-only `Qwen/Qwen2.5-0.5B-Instruct` override for the existing fixtures. The normal run passed both distributed tests and ten capture/packing tests; three existing fixtures could not access the gated Llama model (HTTP 401). The committed default model is unchanged.
- All 24 current feature files differing from merged main were SHA-256 matched to the BJ source tree.
- Ruff configured lint/format, isort, whitespace, and registered-test validation passed. Current-head CI is complete: 106 passed, 25 failed, 45 skipped, 2 cancelled. CUDA Base passed after one failed-jobs-only retry; the NPU reference-image retry passed. Remaining failures are the separate MLX fixture issue, shared AMD/NVFP4 failures, Xeon package/test mismatch, and NPU Wan timing plus dependent-job failures; no confirmed sampling-mask regression remains. Monitoring is paused.

Run the distributed tests with:

```sh
PYTHONPATH=python python3 test/registered/sampling/test_sampling_mask.py TestDistributedSamplingMask -v
```

Earlier verification also covered FlashInfer/PyTorch HTTP requests with overlap on/off, seeded parity, and two live Mooncake P/D cases (disabled/error and enabled metadata handoff).

Upstream main `5177a3ec08` remains merged through `92774ddb48`; no history was rebased. `spec/` remains removed.

Review follow-up commits:
- `90eab32d48`: defer abort release until pending KV offload completes; preserve the optional backend hook and leave base classes unchanged.
- `084b80efce`: reuse prefill abort retirement and preserve `err_type` despite sender exceptions.
- `520e7f7541`: reject the retired environment variable at startup.
- `f79c5a9466`: revert the unrelated MLX test repair. That fix is split into #38739; main still reproduces its two shutdown-fixture failures, so MLX CI may remain red here until the separate repair lands.
- `1d542447e5`: live TP2/PP2 coverage.
- `d73bc1bd7f`: merge distributed coverage into the existing sampling-mask test file; use the two-GPU CUDA runner and preserve the one-GPU AMD suite with only the distributed class skipped.

The earlier three review fixes remain separate commits: `e511bfc998` (opt-in PD metadata), `6f53732040` (bulk CPU copy), and `a723a1a55a` (capture-only helper allocation).

## Speed Tests and Profiling

Qwen3-8B BF16, TP1 on one NVIDIA H200, with FlashInfer sampling and attention and CUDA graphs enabled. Each request uses 128 input tokens, 512 output tokens, `temperature=1`, `top_k=4096`, `top_p=1`, and `return_sampling_mask=true`. The PR uses `--sampling-mask-max-tokens=8192` to accommodate cutoff ties; observed masks contained 4096–4396 tokens.

**Before:** upstream `756d0e0a85`, overlap disabled. **After:** PR `cb30f7c6a9`, overlap enabled. This measures the combined implementation and scheduling change.

Median steady decode throughput from three measured runs after one warmup per configuration; parentheses show the run range.

| Batch size | Before (tokens/s) | After (tokens/s) | Change |
| ---: | ---: | ---: | ---: |
| 1 | 155.47 (154.90–155.56) | 182.37 (181.99–182.40) | +17.30% |
| 16 | 1659.03 (1650.52–1665.77) | 2216.61 (2216.46–2219.79) | +33.61% |
| 64 | 3179.15 (3174.73–3201.97) | 4826.58 (4637.89–4837.76) | +51.82% |

Throughput uses scheduler decode logs over full-batch intervals. Prefill, warmup, and final HTTP response serialization are excluded; sampling-mask capture, copying, and decode result processing are included. All 18 measured trials and six warmups completed with aligned masks/logprobs and no overflow or invalid-result failures.

[Reproducer, exact commands, and full per-run evidence](https://github.com/sgl-project/sglang/blob/76bac722f43640f004b7d1083e175940891ad227/spec/evidence/feature-00-sampling-mask.md#qwen3-8b-decode-throughput).

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #34503593751](https://github.com/sgl-project/sglang/actions/runs/34503593751)
Latest PR Test (Extra): :white_check_mark: [Run #35022699343](https://github.com/sgl-project/sglang/actions/runs/35022699343)
Latest PR Test (AMD ROCm 10): :heavy_minus_sign: **No AMD PR run found for this commit**.

Co-authored-by: Nan Jiang <59716405+nanjiangwill@users.noreply.github.com>

### CODE REVIEW

**Comment by ByronHsu** [comment]
/rerun-test test/registered/sampling/test_sampling_mask.py test/registered/expert_pack/test_expert_pack_mxfp4.py

File: python/sglang/srt/disaggregation/utils.py [hunk 1]
```diff
@@ -349,19 +344,15 @@ def __init__(
...
-                self.output_token_sampling_mask_len = torch.zeros(
-                    (size, 16), dtype=torch.int32, device=device
-                )
-                self.output_token_sampling_mask_idx = torch.zeros(
-                    (size, max_sampling_mask_tokens), dtype=torch.int32, device=device
-                )
-                self.output_token_sampling_logprobs = torch.zeros(
-                    (size, 16), dtype=torch.float32, device=device
-                )
+            self.output_token_sampling_mask_len = torch.zeros(
```
Comment (nanjiangwill) [inline python/sglang/srt/disaggregation/utils.py]: Could we keep sampling-mask metadata opt-in for disaggregated serving? This now allocates a `[size, max_sampling_mask_tokens]` buffer and includes it in every auxiliary handoff, even when no request asks for a sampling mask. At the default cap of 4096, that is 16 KiB per metadata slot and per request handoff. Replacing the environment variable with a server argument makes sense, but the unused path should retain near-zero overhead.

File: python/sglang/srt/managers/scheduler_components/batch_result_processor.py [hunk 2]
```diff
@@ -1123,6 +1170,86 @@ def add_sampling_mask_return_values(
...
+        status_by_batch = [None] * batch_size
+        packed_width = sampling_output.token_ids.shape[1]
+        for row, batch_index in enumerate(batch_indices):
+            status = int(statuses[row])
+            length = int(lengths[row])
+            if status == SamplingMaskStatus.OK and not (0 <= length <= packed_width):
+                status = SamplingMaskStatus.INVALID
+            status_by_batch[batch_index] = status
+            if status == SamplingMaskStatus.OK:
+                masks[batch_index] = sampling_output.token_ids[row, :length].tolist()
```
Comment (nanjiangwill) [inline python/sglang/srt/managers/scheduler_components/batch_result_processor.py]: Could we move the packed token IDs to CPU once before this loop for the non-overlap path? With overlap enabled, `copy_to_cpu()` has already done that; with overlap disabled, this tensor is still on CUDA, so each `.tolist()` can introduce a separate blocking device-to-host synchronization. One bulk copy followed by CPU slicing would preserve the intended synchronization shape.

File: python/sglang/srt/layers/sampler.py [hunk 3]
```diff
@@ -302,38 +319,19 @@ def _sample_from_probs(
...
-                )
-            capture_rows = torch.tensor(
-                capture_rows_list, device=probs.device, dtype=torch.long
-            )
-            capture_all_rows = capture_rows_list == list(range(probs.shape[0]))
-
-        def select_capture_rows(tensor: torch.Tensor) -> torch.Tensor:
-            assert capture_rows is not None
-            return tensor if capture_all_rows else tensor.index_select(0, capture_rows)
+        select_capture_rows = partial(
```
Comment (nanjiangwill) [inline python/sglang/srt/layers/sampler.py]: Small hot-path concern: this `partial` is allocated on every non-greedy sampling step, including the common case where no row requests a mask. Could we call `_select_sampling_mask_rows(...)` directly inside the capture-only branches, or construct this only when `capture_rows is not None`? That keeps the default serving path unchanged.

**Review by nanjiangwill — COMMENTED** [review]
Thanks for carrying the overlap path through the existing async device-to-host result flow. The core direction looks sound, and failing before token commit is the right overflow contract. I left three comments about keeping the unused/default path cheap.

This will also need a semantic rebase onto current main: server feature fields now live in `ExecFeatures`, and the sampling-mask support added in #38279 should be preserved. After rebasing, I would also validate one multi-GPU TP run and one live PP or P/D path, since the TP1 benchmark does not exercise the new collectives or transport.

File: python/sglang/srt/managers/scheduler_components/batch_result_processor.py [hunk 4]
```diff
@@ -1123,6 +1169,87 @@ def add_sampling_mask_return_values(
...
+        if req.multimodal_inputs is not None and req.session is None:
+            req.multimodal_inputs.release_features()
+        if get_memory().enable_hisparse:
+            self.hisparse_coordinator.request_finished(req)
+        prepare_release = getattr(
+            self.model_worker, "prepare_for_kv_cache_release", None
+        )
+        if callable(prepare_release):
+            prepare_release(req)
+        release_kv_cache(req, self.tree_cache, is_insert=False)
```
Comment (nanjiangwill) [inline python/sglang/srt/managers/scheduler_components/batch_result_processor.py]: Could this follow the existing decode-offload release path? A previous successful step may still have an asynchronous KV offload reading this request’s slots, and releasing them here makes those slots reusable before that transfer finishes. `decode_offload_manager.finalize_release_on_finish(req)` already defers release through `offload_inflight`; I think sampling-mask aborts need that branch as well, while retaining `is_insert=False` for the ordinary path.

File: python/sglang/srt/disaggregation/prefill.py [hunk 5]
```diff
@@ -783,6 +787,39 @@ def advance_logprob_pt(i: int, req: Req) -> None:
...
+                    self.clear_pending_chunk_send(req)
+                    prepare_abort(
+                        req,
+                        sampling_mask_finish_reason.message,
+                        status_code=sampling_mask_finish_reason.status_code,
+                    )
+                    req.time_stats.trace_ctx.abort(
+                        abort_info={"reason": sampling_mask_finish_reason.message}
+                    )
+                    req.disagg_kv_sender.abort()
```
Comment (nanjiangwill) [inline python/sglang/srt/disaggregation/prefill.py]: Could we route this through `_retire_aborted_prefill_result()`? This block duplicates that request-ownership cleanup, but calls `disagg_kv_sender.abort()` without the helper’s best-effort guard, so a transport exception can interrupt metadata/KV release. Assigning `req.to_finish = sampling_mask_finish_reason` and using the existing helper would also preserve the finish reason’s `err_type`, which `prepare_abort()` currently drops.

File: python/sglang/srt/environ.py [hunk 6]
```diff
@@ -666,6 +666,8 @@ class Envs:
     # computed dynamically at runtime based on cpu_count; see disaggregation backends.
     SGLANG_DISAGGREGATION_THREAD_POOL_SIZE = EnvInt(None)
     SGLANG_DISAGGREGATION_QUEUE_SIZE = EnvInt(4)
+    # Enable on both P and D with the same --sampling-mask-max-tokens value.
+    SGLANG_ENABLE_DISAGG_SAMPLING_MASK = EnvBool(False)
```
Comment (nanjiangwill) [inline python/sglang/srt/environ.py:670]: The boolean plus the shared server-side cap is a reasonable simplification, but this silently stops honoring the already-merged `SGLANG_DISAGGREGATION_SAMPLING_MASK_MAX_TOKENS` setting. Could we either retain it as a compatibility/deprecation path or fail startup with an explicit migration message when it is set? Silently disabling masks for an existing deployment would be difficult to diagnose.

**Review by nanjiangwill — COMMENTED** [review]
Thanks—the three earlier performance comments are addressed, and removing the spec/ tree helped. The remaining inline comments are lifecycle/configuration issues I think should be resolved before merge. Two cleanup and verification items: 4964140a64 is an unrelated MLX test repair and should be split out, and the new distributed paths still need one real TP2 sampling-mask run plus one live PP run. The Mooncake P/D run is useful, but TP1 and unit-level PP reconstruction do not exercise the status collective or actual PP send/recv.

File: test/registered/sampling/test_sampling_mask.py [hunk 7]
```diff
@@ -24,7 +31,7 @@
     popen_launch_server,
 )
 
-register_cuda_ci(est_time=139, stage="base-b", runner_config="1-gpu-small")
+register_cuda_ci(est_time=250, stage="base-b", runner_config="2-gpu-large")
```
Comment (nanjiangwill) [inline test/registered/sampling/test_sampling_mask.py:34]: Could we keep this file on `1-gpu-small` and move only `TestDistributedSamplingMask` into a focused two-GPU test file? These cases genuinely need a different CI contract, which is the explicit new-file exception in `unit-test-admission.md`. In this CI run the distributed cases took about 136 seconds, while the remaining roughly 119 seconds held a second large GPU unnecessarily. Splitting them preserves the existing suite’s cost and makes the two-GPU requirement explicit.

File: test/registered/unit/disaggregation/test_specv2_kvcache_offloading.py [hunk 8]
```diff
@@ -440,5 +446,58 @@ def test_dropped_req_does_not_pin_offload_state(self):
...
+                        HiCacheAck(None, _FinishedEvent(), [1])
+                    ]
+                    manager._trigger_backup = MagicMock(return_value="hash")
+
+                with (
+                    get_context().override_server_args(
+                        disaggregation_decode_enable_offload_kvcache=True,
+                        enable_hisparse=False,
+                    ),
+                    patch(
```
Comment (nanjiangwill) [inline test/registered/unit/disaggregation/test_specv2_kvcache_offloading.py]: This patch appears unused: with `disaggregation_decode_enable_offload_kvcache=True`, `_handle_sampling_mask_abort()` calls `finalize_release_on_finish()` and never reaches this module’s `release_kv_cache`. Could we remove the patch, and the `model_worker` mock if it is then unused, so the test models only the dependencies it exercises?

**Review by nanjiangwill — COMMENTED** [review]
The lifecycle and configuration fixes look good. I left two focused test-structure comments to keep the added coverage aligned with SGLang’s CI contract and remove an unused mock.

File: python/sglang/srt/arg_groups/validation_hook.py [hunk 9]
```diff
@@ -438,6 +439,22 @@ def validate_prefill_decode_interval(server_args: Any):
         raise ValueError("--prefill-decode-interval must be non-negative.")
 
 
+def validate_sampling_mask_max_tokens(server_args: Any):
+    if envs.SGLANG_DISAGGREGATION_SAMPLING_MASK_MAX_TOKENS.is_set():
```
Comment (nanjiangwill) [inline python/sglang/srt/arg_groups/validation_hook.py:443]: maybe we have a follow up pr to deprecate `SGLANG_DISAGGREGATION_SAMPLING_MASK_MAX_TOKENS` entirely in sglang, but merge it later

**Review by nanjiangwill — APPROVED** [review]
